### ISSUE TITLE
Finer-grained validation tools would be useful

### ISSUE THREAD

**Issue by cormullion** [JuliaData/CSV.jl#225]
Here's a recent experience with CSV.jl which I thought I would share.

tl;dr It's really good, but some more detailed validation information would be very useful, and I think it would be relatively easy to add.

So I was making a not-too-big DataFrame and wrote it to a CSV file successfully. But when I tried to read it in with `CSV.read()`, I saw this error:

    ERROR: ArgumentError: Invalid DateTime string

(One of the DataFrame's columns had been a DateTime column.) So I tried using `CSV.validate()`, but the error was still there:

    ERROR: ArgumentError: Invalid DateTime string

So I read through a few hundred lines of the CSV file, looking at date after date (speed dating!) but couldn't see anything. I ran a few operations on the DataFrame column to check; I couldn't find any problems.

Then I had an idea. This is Julia, so:

```julia
@edit CSV.validate("/tmp/data.csv")
```

and I added a couple of lines to the code:

```julia
function validate(s::CSV.Source)
    sch = Data.schema(s) # size, header, types
    rows, cols = size(sch)
    types = Data.types(sch)
    state = P()
    for row = 1:rows
        rowstr = ""
        print("[$row] ")                                # <---------
        for col = 1:cols
            print(" $col ")                              # <---------
            ...
``` 
   
So now when I run `CSV.validate()` again, I see this:

```
[1]  1  2  3  4  5  6 [2]  1  2  3  4  5  6 ...
```

(t's pretty boring to look at, but...)

```
...[30485]  1  2  3  4  5  6 [30486]  1  2  3  4  5  6 [30487]  1  2  3  4  5  6 [30488]  
1  2  3  4  5  6 [30489]  1  2  3  4  5  6 [30490]  1  2  3  4  5  6 [30491]  1  2  3  4  5  6 
[30492]  1  2  3  4  5  6 [30493]  1  2  3  4  5  6 [30494]  1  2  3  
4 ERROR: ArgumentError: Invalid DateTime string
```

and it was on line 30494 that I found the error. I wouldn't have made it that far into the file any other way.

And, after all that, it wasn't a DateTime-related error at all, but a backslash-quote (`\"`) situation in the preceding column.

So I'm thinking that it would be useful to know on which line the error was found. I'm sure there's a better way to do it than to print out row and column numbers, but that would be better than nothing. Perhaps a `verbose=true` option?

**Comment by nalimilan** [comment]
Yes, for all errors it would be nice to print information on the row (and column if possible) where it happened. But better not print anything about rows which were parsed successfully.

**Comment by cormullion** [comment]
I don't understand how the magic works inside, but my thinking was that perhaps the parsing failure handling was somehow unable to access the current row number ... in which case better to overreport success than underreport failure. But perhaps the parse failure code _can_ access the current row number, in which case I'd agree.

**Comment by quinnj** [comment]
I've been slowly building up a much more robust underlying parsing library: https://github.com/JuliaData/Parsers.jl. This will enable much more useful error reporting and detailed information during CSV.validate. It's actually getting pretty close, so maybe over the next week I can work on switching CSV.jl over to using it.

**Comment by cormullion** [comment]
@quinnj Sounds cool!

**Comment by ExpandingMan** [comment]
I was thinking of opening another issue; right now we have pretty bad error output for when whacky things happen unexpectedly in the midst of a file.  For example, we had people sending us these files which were not supposed to have any missings but there were a whole bunch of them in the middle of a huge file output as `"MISSING"`.  This resulted in an inscrutable error.  It's a fairly usual pattern.  On the bright side CSV.jl itself continues to seem very robust (great work all!), it's just very hard to figure out where legitimate errors are coming from.

**Comment by kescobo** [comment]
I wonder if it would be worth tying into the logging utilities built into the language now - the line/row reporting you're talking about could be `@debug` so that under normal circumstances it would not be show (or even evaluated), but if someone sets the log level to debug they'd get the necessary info (in whatever form that takes). I'm reasonably certain this can complimentary to `verbose=true` (eg that keyword could set `log_level` to `debug` within the function, but I could also set the `log_level` somewhere else in my code and get the same output).

**Comment by quinnj** [comment]
Yeah, that's a great idea.

**Comment by ExpandingMan** [comment]
I like the idea of using the new logging capabilities, but it's hard for me to imagine how a global value could be used to give useful output.  You can't just have it spit something out on every row (well you could, but it would only be useful for very small tables).  I suppose you could have it behave differently when errors are encountered, but the better alternative would be to just have better errors in the first place.  Unless you have something clever in mind that I'm missing?
